src/shape.rs

```diff
@@ -660,237 +676,215 @@ impl ShapeLine {
-                                    ));
+                                    );
+                                    visual_lines.push(current_visual_line);
+                                    current_visual_line = VisualLine::default();
+
                                     number_of_blanks = 0;
-                                    fit_x = line_width - glyph_size;
-                                    word_range_width = glyph_size;
+                                    fit_x = line_width - glyph_width;
+                                    word_range_width = glyph_width;
                                     fitting_start = (i, glyph_i);
                                 }
                             }
                         } else {
                             // Wrap::Word
-                            let mut prev_word_width = None;
-                            if word.blank && number_of_blanks > 0 {
+                            let mut trailing_space_width = None;
+                            if word.blank {
                                 // current word causing a wrap is a space so we ignore it
-                                number_of_blanks -= 1;
-                            } else if let Some(previous_word) = span.words.get(i - 1) {
-                                // Current word causing a wrap is not whitespace, so we ignore the
-                                // previous word if it's a whitespace
-                                if previous_word.blank {
-                                    number_of_blanks -= 1;
-                                    prev_word_width = Some(previous_word.x_advance * font_size);
+                                // number_of_blanks = number_of_blanks.saturating_sub(1);
+                            } else if i > 0 {
+                                if let Some(previous_word) = span.words.get(i - 1) {
+                                    // Current word causing a wrap is not whitespace, so we ignore the
+                                    // previous word if it's a whitespace
+                                    if previous_word.blank {
+                                        trailing_space_width =
+                                            Some(previous_word.x_advance * font_size);
+                                        number_of_blanks = number_of_blanks.saturating_sub(1);
+                                    }
                                 }
                             }
-                            if let Some(width) = prev_word_width {
-                                word_ranges.push((
+                            if let Some(width) = trailing_space_width {
+                                add_to_visual_line(
+                                    &mut current_visual_line,
+                                    span_index,
                                     fitting_start,
-                                    (i - 1, 0),
+                                    (i, 0),
```

Comment:
```suggestion
                                    (i - 1, 0),
```
This handles the case when the last word that fits in the line was blank, right? Previously, the `i - 1` made sure to skip the blank word. In the incongruent case the blank word is skipped by using `i + 2` as start index, so it makes sense that it should also be skipped in the congruent case by using `i - 1` as end index. Or is there another reason this was changed?

Reply:
Since I don't remember why I changed it and your reasoning is sound I'll revert it.